modules/xphoto/src/inpainting.cpp

```diff
@@ -309,96 +1134,190 @@ namespace xphoto
+                        inpaint <float,  3>( src, mask, dst, algorithmType );
+                        break;
+                    case CV_32FC4:
+                        inpaint <float,  4>( src, mask, dst, algorithmType );
+                        break;
+                    case CV_64FC1:
+                        inpaint <double, 1>( src, mask, dst, algorithmType );
+                        break;
+                    case CV_64FC2:
+                        inpaint <double, 2>( src, mask, dst, algorithmType );
+                        break;
+                    case CV_64FC3:
+                        inpaint <double, 3>( src, mask, dst, algorithmType );
+                        break;
+                    case CV_64FC4:
+                        inpaint <double, 4>( src, mask, dst, algorithmType );
+                        break;
+                    default:
+                        CV_Error_( CV_StsNotImplemented,
+                            ("Unsupported source image format (=%d)",
+                            src.type()) );
+                        break;
+                }
                 break;
-            case CV_64FC4:
-                inpaint <double, 4>( src, mask, dst, algorithmType );
+            case xphoto::INPAINT_FSR_BEST:
+            case xphoto::INPAINT_FSR_FAST:
+                CV_Assert( src.channels() == 1 || src.channels() == 3 );
+                double minRange, maxRange;
+                switch ( src.type() )
+                {
+                    case CV_8UC1:
+                        break;
+                    case CV_8UC3:
+                        break;
+                    case CV_16UC1:
+                        cv::minMaxLoc(src, &minRange, &maxRange);
+                        if (minRange < 0 || maxRange > 65535)
+                        {
+                            CV_Error(cv::Error::StsUnsupportedFormat, "Unsupported source image format!");
+                            break;
+                        }
+                        src.convertTo(src, CV_8UC1, 1/257.0);
+                        break;
+                    case CV_32FC1:
+                    case CV_64FC1:
+                        cv::minMaxLoc(src, &minRange, &maxRange);
+                        if (minRange < 0 || maxRange > 1)
+                        {
+                            CV_Error(cv::Error::StsUnsupportedFormat, "Unsupported source image format!");
+                            break;
+                        }
+                        src.convertTo(src, CV_8UC1, 255.0);
+                        break;
+                    case CV_16UC3:
+                        cv::minMaxLoc(src, &minRange, &maxRange);
+                        if (minRange < 0 || maxRange > 65535)
+                        {
+                            CV_Error(cv::Error::StsUnsupportedFormat, "Unsupported source image format!");
+                            break;
+                        }
+                        src.convertTo(src, CV_8UC3, 1 / 257.0);
```

Comment:
BTW, `.convertTo()` does not care about number of channels. `CV_8UC3` can be safely replaced by `CV_8UC1`, or `CV_8U`. Also this case code block can be merged with `case CV_16UC1:` (similar for others).